### PR TITLE
Implement websockets and add a market maker bot

### PR DESCRIPTION
**Pull request by macalinao**

### CODE REVIEW

File: ucf_exchange_client/app.py [hunk 1]
```diff
@@ -21,16 +22,15 @@ def __init__(self, name, default_handlers=True):
...
-                sock.write(msg_to_json(msg._asdict()))
+        self.conn = connection.create(host, port)
+        asyncio.get_event_loop().run_until_complete(self._handler())
+
+    async def _handler(self):
+        """Handles incoming messages and responds to them."""
+        while True:
+            msg = await self.conn.read()
+            resp = await self._handle(msg)
+            for part in resp: self.conn.write(part)
```
Comment (neilvyas) [inline ucf_exchange_client/app.py]: newline pls

File: ucf_exchange_client/app.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 from collections import defaultdict
-from .utils.socket import get_msg, get_conn
-from .RPC_types import msg_from_json, msg_to_json
+import asyncio
```
Comment (neilvyas) [inline ucf_exchange_client/app.py:1]: alphabetize import pls

File: ucf_exchange_client/connection.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import asyncio
+import websockets
+
+from .RPC_types import msg_from_json, msg_to_json
```
Comment (neilvyas) [inline ucf_exchange_client/connection.py]: so I just changed this functionality on `master`, you should `import json` and `json.dumps` the message yourself.

File: examples/market_maker.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+@market_maker.handle('Trade')
+def update_values(state, msg):
+    if msg.ticker not in state.theos:
+        state.theos[msg.ticker] = msg.price
+    else:
+        state.theos[msg.ticker] = (state.theos[msg.ticker] + msg.price) / 2
+    return []
+
+
+market_maker.run('localhost', 8166)
```
Comment (neilvyas) [inline examples/market_maker.py]: does this actually work?

File: examples/market_maker.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+from ucf_exchange_client import Strategy
+
+market_maker = Strategy('mm')
```
Comment (neilvyas) [inline examples/market_maker.py:21]: So as-written this strategy doesn't actually issue any trades. You should define a method that handles `Book` or something and actually does like `yield Buy(ticker, price, quantity)`.

File: examples/market_maker.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+from ucf_exchange_client import Strategy
```
Comment (neilvyas) [inline examples/market_maker.py]: you should also import `Buy`, `Sell`, `Cancel`.

File: examples/market_maker.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+market_maker = Strategy('mm')
+
+@market_maker.handle('Hello')
```
Comment (neilvyas) [inline examples/market_maker.py]: This is now called `'Handshake`.

File: ucf_exchange_client/app.py [hunk 8]
```diff
@@ -52,7 +53,7 @@ def decorator(handler):
             self._add_handler(msg_type, handler)
             return handler
         return decorator
-    
+
```
Comment (neilvyas) [inline ucf_exchange_client/app.py:56]: one newline between method definitions pls

File: ucf_exchange_client/connection.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
+import asyncio
+import json
+import websockets
+
+from .RPC_types import msg_from_json, msg_to_json
+
+
+class Connection:
+    """The connection to the UCF exchange backend."""
```
Comment (neilvyas) [inline ucf_exchange_client/connection.py]: s/backend/server

**Review by neilvyas — CHANGES_REQUESTED** [review]
Pretty good. Need to actually flesh out the market maker bot. `async` code looks good but maybe needs tests? I don't have `asyncio` experience.

File: examples/market_maker.py (hunk 4 above)
Comment (macalinao) [reply examples/market_maker.py]: In theory. I don't know since we don't have a server.

File: ucf_exchange_client/app.py (hunk 8 above)
Comment (macalinao) [reply ucf_exchange_client/app.py:56]: there is. You had trailing whitespace here.

File: ucf_exchange_client/connection.py (hunk 9 above)
Comment (macalinao) [reply ucf_exchange_client/connection.py]: Done.

File: examples/market_maker.py (hunk 5 above)
Comment (macalinao) [reply examples/market_maker.py:21]: ah, I thought this was complete. I'm just going to port my old market maker bot to this API.

File: examples/market_maker.py (hunk 6 above)
Comment (macalinao) [reply examples/market_maker.py]: Ack.

File: examples/market_maker.py (hunk 7 above)
Comment (macalinao) [reply examples/market_maker.py]: Done.

**Comment by macalinao** [comment]
@neilvyas I don't think unit tests in the I/O code are worth it. We should just do an integration test.

File: ucf_exchange_client/app.py (hunk 8 above)
Comment (neilvyas) [reply ucf_exchange_client/app.py:56]: right, thanks. `vim` should have picked that up...

File: examples/market_maker.py (hunk 4 above)
Comment (neilvyas) [reply examples/market_maker.py]: I think I meant this more as a rhetorical question. in any case, you should import-guard this:

``` python
if __name__ == "__main__":
  market_maker.run('localhost', 8166)
```

Comment (macalinao) [reply examples/market_maker.py]: good point my friend

File: ucf_exchange_client/app.py (hunk 8 above)
Comment (macalinao) [reply ucf_exchange_client/app.py:56]: `M-x delete-trailing-whitespace`

**Comment by neilvyas** [comment]
yeah, I will merge without coverage and we can add coverage later. I'd like to get the changes to `Strategy` in before we get hella merge conflicts.

File: ucf_exchange_client/app.py [hunk 10]
```diff
@@ -14,23 +15,23 @@ def __init__(self, name, default_handlers=True):
...
-                # handle the orderid stuff automatically.
-                sock.write(msg_to_json(msg._asdict()))
+        asyncio.get_event_loop().run_until_complete(self._run(host, port))
+
+    async def _run(self, host, port):
+        """Sets up the connection and handles incoming messages."""
+        self.conn = await connection.create(host, port)
+        while True:
+            msg = await self.conn.read()
+            resp = await self._handle(msg)
```
Comment (neilvyas) [inline ucf_exchange_client/app.py]: I don't think you need to await this? It's not an async operation, and `_handle` isn't async-def'd anyway.

**Review by neilvyas — CHANGES_REQUESTED** [review]
i think this is broken

File: ucf_exchange_client/app.py (hunk 10 above)
Comment (macalinao) [reply ucf_exchange_client/app.py]: Done.

**Comment by neilvyas** [comment]
ok, will merge as soon as mm bot is ported.

**Pull request merged**